library/core/src/iter/adapters/zip.rs

```diff
@@ -471,10 +514,21 @@ impl<A: Debug + TrustedRandomAccess, B: Debug + TrustedRandomAccess> ZipFmt<A, B
 ///   `self.__iterator_get_unchecked(idx)`, assuming that the required traits are implemented.
 /// * It must also be safe to drop `self` after calling `self.__iterator_get_unchecked(idx)`.
 /// * If `T` is a subtype of `Self`, then it must be safe to coerce `self` to `T`.
+//
+// FIXME: Clarify interaction with SourceIter/InPlaceIterable. Calling `SouceIter::as_inner`
+// after `__iterator_get_unchecked` is supposed to be allowed.
+#[doc(hidden)]
+#[unstable(feature = "trusted_random_access", issue = "none")]
+#[rustc_specialization_trait]
+pub unsafe trait TrustedRandomAccess: TrustedRandomAccessNoCoerce {}
+
+/// Like [`TrustedRandomAccess`] but without any of the requirements / guarantees around
+/// coercions to subtypes after `__iterator_get_unchecked` (they aren’t allowed here!), and
+/// without the requirement that subtypes / supertypes implement [`TrustedRandomAccessNoCoerce`].
 #[doc(hidden)]
 #[unstable(feature = "trusted_random_access", issue = "none")]
 #[rustc_specialization_trait]
-pub unsafe trait TrustedRandomAccess: Sized {
+pub unsafe trait TrustedRandomAccessNoCoerce: Sized {
```

Comment:
This should live somewhere under `iter::traits` instead of `iter::adapters` since it's useful for other specializations too.

And I'm still thinking about whether they're semantically too close to each other. There are a lot of requirements on `TrustedRandomAccess` that are irrelevant for forward-pass, iterator-consuming loops. E.g. the implications about mixing it with `next_back()` or interactions with cloning the iterator.

And a `TrustedUsizeLen` could be extracted from it too. There are places where that would useful without having indexed access. But perhaps that's too far for this PR.

Reply:
Sure, I agree that there’s lots of potential for re-structuring and optimizing `TrustedRandomAccess` and potential related traits. This PR is mostly focused on keeping the unsoundness fixes of https://github.com/rust-lang/rust/pull/85874 while avoiding any relevant performance regressions. So IMO it would be a good idea to review in just that context.

Also regarding the position, IMO `TrustedRandomAccess` and `TrustedRandomAccessNoCoerce` are so similar that they shouldn’t live in separate places. `TrustedRandomAccess` already was in a weird position before this PR, judging by its additional use in `SpecInPlaceCollect` that makes it no longer `Zip`-specific.

Reply:
But … well, by all means, feel free to open a PR on [`steffahn:fix_unsound_zip_optimization_plus_new_trait`](https://github.com/steffahn/rust/tree/fix_unsound_zip_optimization_plus_new_trait) that moves both traits to another place, if you feel like doing that right now ;-)